src/i18n.ts

```diff
@@ -0,0 +1,342 @@
+	function mapMessages<T extends Messages>(modules: LocaleModule<T>[]): T[] {
+		return modules.map(function (localeModule: LocaleModule<T>): T {
+			return localeModule.default as T;
+		});
+	}
+
+	if (has('host-node')) {
+		return function<T extends Messages>(paths: string[]): Promise<T[]> {
+			const modules = paths.map(function (path: string): LocaleModule<T> {
+				return global.require(path) as LocaleModule<T>;
+			});
+
+			return Promise.all(mapMessages(modules));
+		};
+	}
+
+	return function<T extends Messages>(paths: string[]): Promise<T[]> {
+		return new Promise<T[]>(function (resolve, reject) {
+			const require = global.require;
+			require.on('error', function (error: Error) {
+				reject(error);
+			});
+			require(paths, function (...modules: LocaleModule<T>[]) {
+				resolve(mapMessages(modules));
+			});
+		});
+	};
+})();
+
+/**
+ * @private
+ * Retrieve a list of supported locales that can provide messages for the specified locale.
+ *
+ * @param locale
+ * The target locale.
+ *
+ * @param supported
+ * The locales that are supported by the bundle.
+ *
+ * @return
+ * A list of supported locales that match the target locale.
+ */
+function getSupportedLocales(locale: string, supported: string[] = []): string[] {
+	const normalized = normalizeLocale(locale);
+	const parts = normalized.split('-');
+	const result: string[] = [];
+
+	let current = parts[0];
+
+	if (supported.indexOf(current) > -1) {
+		result.push(current);
+	}
+
+	for (let i = 0; i < parts.length - 1; i += 1) {
+		const next = parts[i + 1];
+
+		if (next) {
+			current += '-' + next;
+
+			if (supported.indexOf(current) > -1) {
+				result.push(current);
+			}
+		}
+	}
+
+	return result;
+}
+
+/**
+ * @private
+ * A type guard that determines whether a context object is stateful and can be invalidated.
+ *
+ * @param value
+ * The target value.
+ */
+function isContextObject(value: any): boolean {
+	return value.state && typeof value.own === 'function' && typeof value.invalidate === 'function';
+}
+
+/**
+ * @private
+ * Normalize a locale so that it can be converted to a bundle path.
+ *
+ * @param locale
+ * The target locale.
+ *
+ * @return The normalized locale.
+ */
+const normalizeLocale = (function () {
+	if (has('host-node')) {
+		return function (locale: string): string {
+			if (locale.indexOf('.') === -1) {
+				return locale;
+			}
+
+			return locale.split('.').slice(0, -1).map(function (part: string): string {
+				return part.replace(/_/g, '-');
+			}).join('-');
+		};
+	}
+
+	return function (locale: string): string {
+		return locale;
+	};
+})();
+
+/**
+ * @private
+ * Extract a locale from the provided context. Either the context's locale or the current locale is returned.
+ *
+ * @param context
+ * An optional context containing the locale information.
+ */
+function resolveLocale(context?: any): string {
+	if (context) {
+		if (typeof context === 'string') {
+			return context;
+		}
+		else if (context.state.locale) {
```

Comment:
A `state` object is required to meet the interface. Are there instances in which a `Stateful` object would not have a `state`?